lib/change_stream.js

```diff
@@ -0,0 +1,227 @@
+ * @param {boolean} [options.disableResume=false] Whether the driver should automatically attempt to resume the change stream in the event of a potentially resumable error (such as a network error)
+ * @param {number} [options.batchSize=null] The number of documents to return per batch. See {@link https://docs.mongodb.com/manual/reference/command/aggregate|aggregation documentation}.
+ * @param {object} [options.collation=null] Specify collation settings for operation. See {@link https://docs.mongodb.com/manual/reference/command/aggregate|aggregation documentation}.
+ * @param {ReadPreference} [options.readPreference=null] The read preference. Defaults to the read preference of the database or collection. See {@link https://docs.mongodb.com/manual/reference/read-preference|read preference documentation}.
+ * @fires ChangeStream#change
+ * @return {ChangeStream} a ChangeStream instance.
+ */
+var ChangeStream = function(changeDomain, pipeline, options) {
+  var self = this;
+  self.changeDomain = changeDomain;
+  self.pipeline = pipeline || [];
+  self.options = options || {};
+
+  // Extract the resume token (if it exists)
+  if (options.resumeAfter) {
+    self.cachedResumeToken = options.resumeAfter;
+  }
+
+  // Extract namespace and serverConfig from changeDomain
+  self.namespace = {};
+  if (changeDomain.collectionName) {
+    self.namespace.database = self.changeDomain.s.db.databaseName;
+    self.namespace.collection = self.changeDomain.collectionName;
+    self.serverConfig = self.changeDomain.s.db.serverConfig;
+  } else if (self.changeDomain.databaseName) {
+    self.namespace.database = self.changeDomain.databaseName;
+    self.serverConfig = self.changeDomain.serverConfig;
+  } else {
+    throw new Error('changeDomain must be either a database or collection.');
+  }
+
+  self.cursor = createChangeStreamCursor(self);
+
+  self.on('newListener', function(eventName) {
+    if (eventName === 'change') {
+      self.changeListenerExists = true;
+      self.cursor.on('data', function (change) {
+        processNewChange(self, null, change, 'eventEmitter');
+      });
+    }
+  });
+
+};
+
+// Create a new change stream cursor based on self's configuration
+var createChangeStreamCursor = function (self) {
+  var changeStreamCursor;
+
+  if (self.cachedResumeToken) {
+    self.options.resumeAfter = self.cachedResumeToken;
+  }
+
+  var serverSupportsChangeNotifications = false;
+  if (serverSupportsChangeNotifications) {
+    var collectionName = self.changeDomain.collectionName ? self.changeDomain.collectionName : 1;
+    changeStreamCursor = buildChangeStreamAggregationCommand(self, collectionName);
+
+  } else {
+    // Generate a no conflict value for this Change Stream if it does not already exist.
+    // This prevents collisions when multiple change stream simulations exist simulataneously.
+    self.options.cursorNoConflict = self.options.cursorNoConflict || Math.random().toString().substring(2, 12);
```

Comment:
also it looks like this thing is used only in one function below, I'm not sure it wants to be in `options`